calm-ingest/calm_to_dynamodb.py

```diff
@@ -0,0 +1,212 @@
+    # just interested in extracting those.
+    context = etree.iterparse(
+        path, events=('end',), tag='DScribeRecord', resolve_entities=False)
+    for entry in context:
+        yield entry[1]
+
+
+def _skip_updated_records(records, since_date=None):
+    """
+    Generate a sequence of Calm records that were updated on or after
+    a given date.
+
+    :param records: A generator of ``xml.etree.Element`` instances.
+    :param since_date: (optional) A ``datetime`` object which specifies when
+        the Calm data was last ingested.  If supplied, only records with
+        a modified date on or after this data will be provided.
+    """
+    # If we don't have a date, return everything
+    if since_date is None:
+        yield from records
+
+    else:
+        for record in records:
+            # The modification/creation date of a record is kept in the
+            # <Modified> and <Created> tags, respectively.
+            #
+            # Some records do not have these tags, or their value is empty:
+            # then we always include the record, just in case.
+            modified = record.find('Modified')
+            if (modified is None) or (modified.text is None):
+                modified_str = record.find('Created').text
+            else:
+                modified_str = modified.text
+
+            # The format of these fields is DD/MM/YYYY.
+            if modified_str is None:
+                yield record
+            else:
+                modified = datetime.strptime(modified_str, '%d/%m/%Y').date()
+                if modified >= since_date:
+                    yield record
+
+
+def _prepare_record_for_dynamodb(records):
+    """
+    Generate a sequence of dictionaries that can be inserted into DynamoDB.
+    """
+    for record in records:
+        # Produce a dict of all the attributes on the record that aren't
+        # stored in one of the other Dynamo fields.  We'll convert this to
+        # a JSON string and store the encoded output, so discarding fields
+        # we already have improves performance and storage.
+        data = {
+            c.tag: c.text
+            for c in record.getchildren()
+            if c.tag not in ('RecordID', 'RecordType', 'RefNo', 'AltRefNo')
+        }
+
+        yield {
+            'RecordID': record.find('RecordID').text,
+            'RecordType': record.find('RecordType').text,
+            'RefNo': record.find('RefNo').text,
+            'AltRefNo': record.find('AltRefNo').text,
+            'data': json.dumps(data),
+        }
+
+
+if __name__ == '__main__':
+    args = parse_args()
+
+    dynamodb = boto3.resource('dynamodb')
+    table = dynamodb.Table('CalmData')
+
+    # We do some bookkeeping to avoid sending the same record to DynamoDB
+    # multiple times, which is both slow and unnecessary cost.  The full
+    # Calm export has ~331k records, so the more we can skip, the better.
+    try:
+        existing = json.load(open('existing.json', 'r'))
```

Comment:
I guess this is ok for now, but loading from a file like this makes parallelisation impossible.